### PR TITLE
Fix archive_write_files on Windows (add zip extract test)

### PR DESCRIPTION
**Pull request by cielavenir**
This temporal pull request simply adds zip version of "can strip components if desired" test, but it raises "4821 in central directory, 4972 in local header" on Windows. It is not related to password, so super-strange.

Could you take a look? @gaborcsardi

### CODE REVIEW

**Comment by cielavenir** [comment]
calling archive_write_close is not sufficient

**Comment by gaborcsardi** [comment]
@cielavenir Thanks, I can take a look later today.

**Comment by cielavenir** [comment]
The thing is, is my library usage is wrong or is there potential problem on this library?

**Comment by cielavenir** [comment]
Hello @gaborcsardi , it turned out that open() argument requires O_BINARY on Windows. I have not checked in detail but it is possible that on Windows archive_write_files would not be working from the beginning.

I added integrity check to avoid this kind of accident further.

**Comment by cielavenir** [comment]
Simply adding integrity check to tar failed as https://github.com/r-lib/archive/actions/runs/4074516161
This means tar content was corrupted.

**Comment by cielavenir** [comment]
now it is ready.

File: src/archive_write.cpp [hunk 1]
```diff
@@ -73,13 +73,15 @@ void rchive_write_close(Rconnection con) {
   struct archive* out;
   struct archive_entry* entry;
   in = archive_read_disk_new();
-#ifndef __MINGW32__
+#if defined(_WIN32) || (!defined(__GNUC__) && !defined(__clang__))
```
Comment (gaborcsardi) [inline src/archive_write.cpp:76]: How is `O_BINARY` related to GCC and clang?

Comment (cielavenir) [reply src/archive_write.cpp]: I just don't trust msvc/bcc so much and am not sure if they correctly add _WIN32, but sure let's trust.

I just mean GCC/clang ensure to add _WIN32.

**Comment by cielavenir** [comment]
hi, is cf0b568 good or still bad

**Comment by gaborcsardi** [comment]
I think it is good.

**Comment by cielavenir** [comment]
I don't think there are remaining issue here

**Comment by cielavenir** [comment]
are there still anything interfering merging?

**Comment by cielavenir** [comment]
This one should be ready, could you merge or are there any issues?

File: src/archive_write.cpp [hunk 2]
```diff
@@ -73,13 +73,15 @@ void rchive_write_close(Rconnection con) {
   struct archive* out;
   struct archive_entry* entry;
   in = archive_read_disk_new();
-#ifndef __MINGW32__
+#if defined(_WIN32)
```
Comment (gaborcsardi) [inline src/archive_write.cpp:76]: Is there a reason for changing this?

File: tests/testthat.R [hunk 3]
```diff
@@ -1,4 +1,5 @@
 library(testthat)
 library(archive)
+library(tools)
```
Comment (gaborcsardi) [inline tests/testthat.R:3]: Please do not load tools here, but qualify references with `::` in the test files.

**Review by gaborcsardi — CHANGES_REQUESTED** [review]

File: tests/testthat.R (hunk 3 above)
Comment (cielavenir) [reply tests/testthat.R]: Fixed; I used R for 15 years but I didn't even know the feature.

File: src/archive_write.cpp (hunk 2 above)
Comment (cielavenir) [reply src/archive_write.cpp:76]: a... maybe please check https://github.com/r-lib/archive/pull/86 first then let's go back to this one

Comment (cielavenir) [reply src/archive_write.cpp]: merged #86 to this one (so that #86 and #80 can be merged by one review round)

File: src/archive_write.cpp (hunk 1 above)
Comment (cielavenir) [reply src/archive_write.cpp]: @gaborcsardi FYI I dropped MSVC/clang support.

**Comment by cielavenir** [comment]
@gaborcsardi now fixing archive_write_files on Windows depends on this PR. Please take a look.

File: src/archive_write.cpp (hunk 2 above)
Comment (cielavenir) [reply src/archive_write.cpp]: @gaborcsardi "Allow edits and access to secrets by maintainers" if on, so feel free to push the minor flag and complete it...

**Comment by cielavenir** [comment]
I'm tired, converted to issue https://github.com/r-lib/archive/issues/98

**Comment by cielavenir** [comment]
alternatively we can use `#ifndef O_BINARY`

**Pull request closed without merging**